client/state/data-layer/extensions-middleware.js

```diff
@@ -1,101 +1,58 @@
 /**
- * External dependencies
- */
-import { toPairs, fromPairs } from 'lodash';
-
-/**
  * Internal dependencies
  */
-import { local, mergeHandlers } from './utils';
-
-const stripHandlersForKey = ( key, keyHandlers, keyHandlersToStrip ) => {
-	const newKeyHandlers = (
-		keyHandlersToStrip
-			? keyHandlers.filter( ( handler ) => ! keyHandlersToStrip.includes( handler ) )
-			: keyHandlers
+import { mergeHandlers } from './utils';
+import { middleware } from './wpcom-api-middleware';
+import debugFactory from 'debug';
+
+const debug = debugFactory( 'calypso:state:data-layer' );
+let extensionHandlers = Object.create( null );
+let handlerMiddleware = buildMiddleware( extensionHandlers );
+
+export function addHandlers( extensionName, handlers, existingHandlers = extensionHandlers ) {
+	if ( Object.keys( existingHandlers ).includes( extensionName ) ) {
+		throw new Error( `Handlers for extension ${ extensionName } already present.` );
```

Comment:
I think @blowery wants to keep exceptions out of the browser. I am not 100% convinced. either way, I think we should belabor of this one here. certain errors in the data-layer are "fine" because they are triggered asynchronously on the return of some event such as a network call. `addHandlers()` on the other hand is likely being called synchronously from within existing code and an exception would abort all of the remaining thread execution.

might it be better to return an error?